File: gradle-scala-kotlin/lib/src/main/scala/com/sapient/learning/Generics.scala
package com.sapient.learning

class InvariantGarden[A]

class CovariantGarden[+A]

class ContravariantGarden[-A]

class Fruit

class Banana extends Fruit

class Apple extends Fruit

class Generics {

  def main(args: Array[String]): Unit = {
    //val myInvariantGarden: InvariantGarden[Fruit] = new InvariantGarden[Banana] // Illegal statement
    //val myInvariantGarden: InvariantGarden[Banana] = new InvariantGarden[Fruit] // Illegal statement

    val myCovariantGarden: CovariantGarden[Fruit] = new CovariantGarden[Banana]
    //val myCovariantGarden: CovariantGarden[Banana] = new CovariantGarden[Fruit] // Illegal statement

    //val myContravariantGarden: ContravariantGarden[Fruit] = new ContravariantGarden[Banana] // Illegal statement
    val myContravariantGarden: ContravariantGarden[Banana] = new ContravariantGarden[Fruit]
  }

}

File: gradle-scala-kotlin/lib/src/main/kotlin/com/kchandrakant/learning/Generics.kt
package com.kchandrakant.learning

class InvariantGarden<T>
class CovariantGarden<out T>
class ContravariantGarden<in T>

open class Fruit

class Apple : Fruit()

class Banana : Fruit()

fun main() {
    //var myInvariantGarden: InvariantGarden<Fruit> = InvariantGarden<Apple>() //This is illegal
    //var myInvariantGarden: InvariantGarden<Apple> = InvariantGarden<Fruit>() //This is illegal

    var myCovariantGarden: CovariantGarden<Fruit> = CovariantGarden<Apple>() //This is legal
    //var myCovariantGarden: CovariantGarden<Apple> = CovariantGarden<Fruit>() //This is illegal

    //var myContravariantGarden: ContravariantGarden<Fruit> = ContravariantGarden<Apple>() //This is legal
    var myContravariantGarden: ContravariantGarden<Apple> = ContravariantGarden<Fruit>() //This is illegal
}